packages/opentelemetry-tracing/src/AbstractBasicTracerFactory.ts

```diff
@@ -0,0 +1,46 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as types from '@opentelemetry/types';
+import { BasicTracer } from './BasicTracer';
+import { SpanProcessor } from './SpanProcessor';
+
+export abstract class AbstractBasicTracerFactory
```

Comment:
Why is the abstract factory specific to basic tracer? Why not just `AbstractTracerFactory` and `_newTracer` would return `types.Tracer`?

`BasicTracerFactory` and `NodeTracerFactory` would inherit from it as `NodeTracer` and `BasicTracer` both implement `types.Tracer`

Reply:
It's due to this line specifically, since span processors is not an API level concept.  https://github.com/open-telemetry/opentelemetry-js/pull/420/files#diff-7de0ceee2be4a0e5c9c8ba090fcc1ce4R28

Reply:
But the base class they inherit from should be more general. `BasicTracer` and `WebTracer` are both just specializations of `types.Tracer`. The abstract base class is just an abstract tracer factory which all other tracer factories inherit from. Those more specific tracer factories would then have more specific typings.

```typescript
class AbstractTracerFactory {
  protected abstract _newTracer(): types.Tracer;
}

class BasicTracerFactory {
  protected _newTracer(): BasicTracer {
    return new BasicTracer();
  }
}

class WebTracerFactory {
  protected _newTracer(): WebTracer {
    return new WebTracer();
  }
}
```

Reply:
Cool, updated the abstract factory to be more general.